qdr/_ranker.cc

```diff
@@ -0,0 +1,416 @@
+
+#include <math.h>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <stdexcept>
+
+
+typedef std::unordered_map<std::string, std::pair<uint64_t, uint64_t> >
+    counts_t;
+
+
+// count of word occurances in a doc
+typedef std::unordered_map<std::string, uint64_t> word_counts_t;
+
+// a tokenized document
+typedef std::vector<std::string> doc_t;
+
+word_counts_t count_words(const doc_t& document)
+{
+    /// Given a vector of tokens, return a map token -> count in vector
+    word_counts_t ret(document.size() * 3);
```

Comment:
I don't understand why you are multiplying the size of the document times 3 here. It seems like the number of unique words has an upper bound on the size of the document itself.